rest_framework/schemas.py

```diff
@@ -256,6 +256,289 @@ def get_allowed_methods(self, callback):
+                    name='data',
+                    location='body',
+                    required=True,
+                    schema=coreschema.Array()
+                )
+            ]
+
+        if not isinstance(serializer, serializers.Serializer):
+            return []
+
+        fields = []
+        for field in serializer.fields.values():
+            if field.read_only or isinstance(field, serializers.HiddenField):
+                continue
+
+            required = field.required and method != 'PATCH'
+            field = coreapi.Field(
+                name=field.field_name,
+                location='form',
+                required=required,
+                schema=field_to_schema(field)
+            )
+            fields.append(field)
+
+        return fields
+
+    def get_pagination_fields(self, path, method):
+        view = self.view
+
+        if not is_list_view(path, method, view):
+            return []
+
+        pagination = getattr(view, 'pagination_class', None)
+        if not pagination:
+            return []
+
+        paginator = view.pagination_class()
+        return paginator.get_schema_fields(view)
+
+    def get_filter_fields(self, path, method):
+        view = self.view
+
+        if not is_list_view(path, method, view):
+            return []
+
+        if not getattr(view, 'filter_backends', None):
+            return []
+
+        fields = []
+        for filter_backend in view.filter_backends:
+            fields += filter_backend().get_schema_fields(view)
+        return fields
+
+    def get_encoding(self, path, method):
+        """
+        Return the 'encoding' parameter to use for a given endpoint.
+        """
+        view = self.view
+
+        # Core API supports the following request encodings over HTTP...
+        supported_media_types = set((
+            'application/json',
+            'application/x-www-form-urlencoded',
+            'multipart/form-data',
+        ))
+        parser_classes = getattr(view, 'parser_classes', [])
+        for parser_class in parser_classes:
+            media_type = getattr(parser_class, 'media_type', None)
+            if media_type in supported_media_types:
+                return media_type
+            # Raw binary uploads are supported with "application/octet-stream"
+            if media_type == '*/*':
+                return 'application/octet-stream'
+
+        return None
+
+# Note: With `AutoSchema` defined we attach it to APIView.
+#   * We do this here to avoid the dependency cycle from SchemaView needing
+#     APIView (below).
+#   * This requires importing _something_ from `rest_framework.schemas` or
+#     `rest_framework.documentation` before `APIView.schema will be available.
+#       * ???: When would `APIView.schema` be needed and that NOT be the case?
+#   * The alternative is to import AutoSchema to `views`, make `schemas` a
+#     package, and move SchemaView to `schema.views`, importing APIView there.
+APIView.schema = AutoSchema()
```

Comment:
I'd *really* like to see us able to put this on the APIView class directly - I think that'd add some clarity. I guess the lowest impact way to do that is to have these new inspectors in `inspectors.py`, with a bit of docstring at the top of `schemas.py` and `inspectors.py` detailing that inspectors is for the view inspection, and schemas is for bringing it all together.

Reply:
OK. No problem. 

The import cycle is just caused by `SchemaView` being an `APIView` subclass. We can restructure and then import `AutoSchema` into `views`.

This is the only reason to break up the file. It's not anywhere near too big to comfortably handle. 

Rather than create a whole load (more) files directly under `rest_framework`, we can make `schemas` a package. Something like: 

```
rest_framework/
  ...
  schemas/
    __init__.py
    inspectors.py
    generators.py
    views.py
```

Then we expose the key items in the API as they are now, so a user would do something like: 

```
from rest_framework.schemas import (
  AutoSchema, 
  ManualSchema, 
  SchemaGenerator, 
  get_schema_view,
)
```

This would be unchanged for anyone who isn't importing `SchemaView` directly (but I can't help them). I would imagine 99% of users being entirely unaffected. 

**Question**: `schema` is already plural, so do we really want `schemas` with the extra `s`? (If so fine, if not a small breaking change to rename now would be appropriate — it would be easy to fix in user-projects.)